base/essentials.jl

```diff
@@ -141,33 +141,6 @@ end
 argtail(x, rest...) = rest
 tail(x::Tuple) = argtail(x...)
 
-# TODO: a better / more infer-able definition would pehaps be
-#   tuple_type_head(T::Type) = fieldtype(T::Type{<:Tuple}, 1)
```

Comment:
I think we should instate this definition of `tuple_type_head`, and add a correct implementation of `tuple_type_tail`.

Reply:
Hmm....what would a correct implementation of `tuple_type_tail` look like when we have covariance + diagonal constraints?

Reply:
Well, you seem content with `fieldtype((Tuple{T,T} where T<:AbstractFloat), 1) == AbstractFloat`, so it might not be a problem.

`tuple_type_tail(Tuple{T,T} where T) == Tuple{T} where T` seems correct to me. Do you have a counterexample?

Reply:
Here's an example that seems wrong to me for master's `tuple_type_tail`:

```julia
julia> t = (Val(Any), 1, 1.0)
(Val{Any}(), 1, 1.0)

julia> typeof(t) <: Tuple{Val{T},T,T} where T
true

julia> typeof(Base.tail(t)) <: Base.tuple_type_tail(Tuple{Val{T},T,T} where T)
false
```

That's what I meant by covariance + diagonal constraints mucking things up, but I could totally be ignorant about the correct meaning of `tuple_type_tail` - apologies if so!

> Well, you seem content with fieldtype((Tuple{T,T} where T<:AbstractFloat), 1) == AbstractFloat, so it might not be a problem.

Yeah, @vtjnash told me `fieldtype` usually gets the right answers in these cases, which is why I replaced all the usages of `tuple_type_head`/`tuple_type_tail` with it. I'm just not sure how to use `fieldtype` to compute `tuple_type_tail` in general e.g. on vatuple types with unconstrained lengths.